cranelift/codegen/src/isa/x64/inst/mod.rs

```diff
@@ -160,6 +160,14 @@ pub enum Inst {
         dst: Writable<Reg>,
     },
 
+    /// Arithmetic SIMD shifts.
+    XmmShift {
+        width: LaneWidth,
+        kind: ShiftKind,
+        src: RegMemImm,
+        dst: Writable<Reg>,
+    },
```

Comment:
We don't necessarily have to things this way--an alternate approach would be to have all of the different opcodes for packed shifts live separately under `SseOpcode`. I only did this to match what I saw for scalar shifts.

Reply:
I agree with that; shifts are not magically different here, so they should just be part of `SseOpcode`.

Reply:
Ah, ignore my previous comment.  In fact vector shifts by immediates do need a separate `Inst` case because they have a different "shape" from normal XMM insns -- the read-only (`src`) operand is an immediate, not a RM.  Note that `RegMemImm` has no place in XMM-world as far as I know.  Hence I suggest the following:
```
/// XMM shifts by immediates.
XmmShiftByImm {
   opcode: SseOpcode,
   amount: u8,
   dst: Writable<Reg>
}
```
Then add the necessary extra entries to `SseOpcode`, and have the emit function panic for `Inst::XmmShiftByImm` if `opcode` is not one of the few acceptable values.

Maybe this should even be called `XMM_I_R` (so as to be consistent with naming `XMM_RM_R`).  Then we could use it for any XMM insn which takes an immediate and a modifiable register.  I can't think of any except the abovementioned shifts though.

Reply:
Agreed here; the "shape" is equivalent to the `InstructionFormat` here, so it makes sense to add a new vcode inst.

tiny nit re: naming, since we're moving to some more Rusty-like naming, I'd rather go for properly camelCased `XmmImmReg` here (there's going to be a PR to make the names more consistent soonish).

Reply:
>  Note that RegMemImm has no place in XMM-world as far as I know

I think in this case it does. Packed shifts accept either an `imm8` or a `xmm/m128` as operands. So there should be a new `Inst` like:

```
... {
  opcode: SseOpcode,
  src: RegMemImm,
  dst: Writable<Reg>
}
```

Right? But what do I call it?

Reply:
Ah, ok.  So the src operand does have to be `RegMemImm`.

As for a name, what it really wants to have is `Xmm_RegMemImm_Reg`.  But the Rust naming rules would require jamming all that together, `XmmRegMemImmReg`, which is ambiguous: the underscores give meaning.  @bnjbvr do you think this is one place where we might apply for an exemption to the normal rules?

Reply:
In Zulip @bnjbvr suggested `XmmRmiReg` which is shorter and avoids the ambiguity.